rx_signal_processing/decimate.cu

```diff
@@ -1,221 +1,391 @@
-    }                                       // REPLY okay. We decided samples will be corrected at the end so this can stay for now
-    else {
-        auto final_offset = channel_offset + dec_sample_offset + threadIdx.x;
-        sample = original_samples[final_offset];
-    }
-
-
-    filter_products[tap_offset] = cuCmulf(sample,filter_taps[tap_offset]); // REVIEW #4 tell user that this comes from cuComplex.h, any side effects?
-                                                                           // REPLY dont think so.
-    __syncthreads(); // REVIEW #1 Synchronizes all threads in a block, meaning 1 output sample per rx freq is ready to be calculated with the parallel reduce
-
-    auto calculated_output_sample = parallel_reduce(filter_products, tap_offset); // REVIEW #26 Should this be called something like 'decimated_sample' instead - to indicate that it is going into the array of decimated samples? total_sum could be the variable name in parallel reduce, but not here in this context
-                                                                   // REPLY alright
-
-    //this is all wrong now that I think about it. At least for second and third stage
-    if (threadIdx.x == 0) { // REVIEW #1 Explain how you're setting up the array of decimated samples
-        channel_offset = channel_num * samples_per_channel/dm_rate; // REVIEW #13 gridDimx is already samples_per_channel/dm_rate, use it instead
-                                                                    // REPLY it is, yet I've been trying to avoid direct use of CUDA builtin types since it not obvious what they mean
-        auto total_channels = blockDim.y; // REVIEW #0 This should be gridDim.y if you intend to use 'total_channels' as antennas (should use 'antennas')
-        auto freq_offset = threadIdx.y * total_channels; // REVIEW #0 still need to multiply by gridDim.x here to get index into proper location
-        auto total_offset = freq_offset + channel_offset + dec_sample_num;
-        decimated_samples[total_offset] = calculated_output_sample;
-    }
+  cuComplex* decimated_samples,
+  cuComplex* filter_taps, uint32_t dm_rate,
+  uint32_t samples_per_antenna) {
+
+  // Since number of filter taps is calculated at runtime and we do not want to hardcode
+  // values, the shared memory can be dynamically initialized at invocation of the kernel.
+  // http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#shared
+
+  extern __shared__ cuComplex filter_products[];
+
+  auto antenna_num = blockIdx.y;
+  auto antenna_offset = antenna_num * samples_per_antenna;
+
+  auto dec_sample_num = blockIdx.x;
+  auto dec_sample_offset = dec_sample_num * dm_rate;
+
+  auto tap_offset = threadIdx.y * blockDim.x + threadIdx.x;
+
+  // If an offset should extend past the length of samples per antenna
+  // then zeroes are used as to not segfault or run into the next buffer.
+  // output samples using these will be discarded as to not introduce edge effects
```

Comment:
This will have to be updated when that is decided. We still need to a plan on how to discard those samples and where to do it. Similarily, we still need to do the phase correction.